File: linkedlist/insertNode.c
#include <stdio.h>
#include <malloc.h>

typedef struct n{
  int value;
  struct n* next;
} node;

//node* head = NULL;
#define NEWNODE malloc(sizeof(node))

void insertnode(node** head, int value, int pos)
{
  int i = 0;
  node* temp = *head;
  node* newnode;

  //empty list, create head
  if (*head == NULL) {
    *head = NEWNODE;
    (*head)->value = value;
    (*head)->next = NULL;
    return;
  }

  //allocate new node and wait to insert at right pos
  newnode = NEWNODE;
  newnode->value = value;

  //if need change head
  if (pos == 0) {
    newnode->next = *head;
    *head = newnode;
    return;
  }

  //traverse to the end or to the position
  for (i = 0 ; temp->next != NULL && i < pos-1 ; i++)
    temp = temp->next;

  //if position is before end
  if (temp->next != NULL) {
    newnode->next = temp->next;
    temp->next = newnode;
  }
  //if position after the end
  else {
    newnode->next = NULL;
    temp->next = newnode;
  }
}

void reverseList(node** head) {
  node* prev = NULL;
  node* curr = *head;
  node* next = NULL;

  for (;curr != NULL;) {
    next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
  }
  *head = prev;
}

void printList(node** head)
{
  node* temp = *head;

  for (;temp->next != NULL;) {
    printf("%d\n", temp->value);
    temp = temp->next;
  }
  printf("%d\n", temp->value);
}

int main(void) {
  node* head = NULL;

  insertnode(&head, 50, 0);
  insertnode(&head, 51, 0);
  insertnode(&head, 52, 0);
  insertnode(&head, 53, 0);
  insertnode(&head, 54, 0);
  insertnode(&head, 55, 0);
  insertnode(&head, 56, 0);
  insertnode(&head, 57, 0);
  insertnode(&head, 40, 0);
  insertnode(&head, 30, 1);
  insertnode(&head, 20, 100);

  printList(&head);
  reverseList(&head);
  printf("\nReversedList\n");
  printList(&head);
  return 0;
}

File: linkedlist/insertNode.py
class Node:
  def __init__(self, value):
    self.value = value
    self.next = None
    #self.prev = None

class LinkedList:
  def __init__(self):
    self.head = None
    #self.tail = None

  def insertNode(self, value):
    newNode = Node(value)

    if self.head is None:
      self.head = newNode
    else:
      temp = self.head;
      while (temp.next):
        temp = temp.next
      temp.next = newNode

  def mergedList(self, ll1, ll2):
    if ll1 is None:
      return ll2
    if ll2 is None:
      return ll1

    if ll1.value < ll2.value:
      self.head = ll1
      ll1 = ll1.next
    else:
      self.head = ll2
      ll2 - ll2.next


  def printlist(self):
    temp = self.head
    while (temp):
      print(temp.value)
      temp = temp.next

ll1 = LinkedList()
ll2 = LinkedList()

ll1.insertNode(5)
ll1.insertNode(7)
ll1.insertNode(9)
ll1.insertNode(11)

ll2.insertNode(6)
ll2.insertNode(8)
ll2.insertNode(10)
ll2.insertNode(12)
ll2.insertNode(14)
ll2.insertNode(16)

ll1.printlist()
print('\n')
ll2.printlist()

ml = LinkedList()
ml.mergedList(ll1, ll2)
